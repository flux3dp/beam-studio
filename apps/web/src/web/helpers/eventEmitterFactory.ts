import EventEmitter from 'eventemitter3';

const eventEmitters = {
  'flux-id': null,
  'top-bar': null,
  'top-bar-hints': null,
  'time-estimation-button': null,
  'top-bar-menu': null,
  'layer-panel': null,
  'right-panel': null,
  'object-panel': null,
  'alert-progress': null,
  'zoom-block': null,
  'beambox-preference': null,
  workarea: null,
  monitor: null,
  canvas: null,
};
export default {
  createEventEmitter: (type?: string): EventEmitter => {
    if (!type) return new EventEmitter();
    if (!eventEmitters[type]) eventEmitters[type] = new EventEmitter();
    return eventEmitters[type];
  },
};
