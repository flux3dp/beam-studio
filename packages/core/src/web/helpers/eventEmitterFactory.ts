import { EventEmitter } from 'eventemitter3';

const eventEmitters: Record<string, any> = {
  'alert-progress': null,
  'beambox-preference': null,
  canvas: null,
  'flux-id': null,
  'layer-panel': null,
  monitor: null,
  'object-panel': null,
  'right-panel': null,
  'time-estimation-button': null,
  'top-bar': null,
  'top-bar-hints': null,
  'top-bar-menu': null,
  workarea: null,
  'zoom-block': null,
};

export default {
  createEventEmitter: (type?: string): EventEmitter => {
    if (!type) {
      return new EventEmitter();
    }

    if (!eventEmitters[type as keyof typeof eventEmitters]) {
      eventEmitters[type as keyof typeof eventEmitters] = new EventEmitter();
    }

    return eventEmitters[type as keyof typeof eventEmitters] as unknown as EventEmitter;
  },
};
