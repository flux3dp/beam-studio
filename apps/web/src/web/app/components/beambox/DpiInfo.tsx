import classNames from 'classnames';
import React, { useEffect, useState } from 'react';

import BeamboxPreference from 'app/actions/beambox/beambox-preference';
import constant from 'app/actions/beambox/constant';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import useI18n from 'helpers/useI18n';
import { useIsMobile } from 'helpers/system-helper';

import styles from './DpiInfo.module.scss';

const eventEmitter = eventEmitterFactory.createEventEmitter('dpi-info');
const { dpiValueMap } = constant;
type TDpiKey = keyof typeof dpiValueMap;

// TODO: Rearrange corner info/buttons together
const DpiInfo = (): JSX.Element => {
  const lang = useI18n().beambox.document_panel;
  const [engraveDpi, setEngraveDpi] = useState<TDpiKey>(BeamboxPreference.read('engrave_dpi'));
  const isMobile = useIsMobile();

  useEffect(() => {
    eventEmitter.on('UPDATE_DPI', setEngraveDpi);
    return () => {
      eventEmitter.off('UPDATE_DPI', setEngraveDpi);
    };
  }, []);

  return (
    <div className={classNames(styles.container, { [styles.mobile]: isMobile })}>
      <i>
        {lang.engrave_dpi}: {constant.dpiValueMap[engraveDpi]} DPI
      </i>
    </div>
  );
};

export default DpiInfo;
