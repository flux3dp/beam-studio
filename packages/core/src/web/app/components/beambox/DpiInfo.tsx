import React from 'react';

import classNames from 'classnames';

import constant from '@core/app/actions/beambox/constant';
import { useBeamboxPreference } from '@core/helpers/hooks/useBeamboxPreference';
import { useIsMobile } from '@core/helpers/system-helper';
import useI18n from '@core/helpers/useI18n';

import styles from './DpiInfo.module.scss';

// TODO: Rearrange corner info/buttons together
const DpiInfo = (): React.JSX.Element => {
  const lang = useI18n().beambox.document_panel;
  const engraveDpi = useBeamboxPreference('engrave_dpi');
  const isMobile = useIsMobile();

  return (
    <div className={classNames(styles.container, { [styles.mobile]: isMobile })}>
      <i>
        {lang.engrave_dpi}: {constant.dpiValueMap[engraveDpi]} DPI
      </i>
    </div>
  );
};

export default DpiInfo;
