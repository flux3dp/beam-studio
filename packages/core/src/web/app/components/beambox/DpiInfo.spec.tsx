import React from 'react';
import { render } from '@testing-library/react';

import eventEmitterFactory from 'helpers/eventEmitterFactory';

import DpiInfo from './DpiInfo';

jest.mock('helpers/useI18n', () => () => ({
  beambox: {
    document_panel: {
      engrave_dpi: 'Resolution',
    },
  },
}));

const mockRead = jest.fn();
jest.mock('app/actions/beambox/beambox-preference', () => ({
  read: () => mockRead(),
}));

const mockUseIsMobile = jest.fn();
jest.mock('helpers/system-helper', () => ({
  useIsMobile: () => mockUseIsMobile(),
}));

describe('test DpiInfo', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {
    const eventEmitter = eventEmitterFactory.createEventEmitter('dpi-info');
    mockRead.mockReturnValue('low');
    const { container } = render(<DpiInfo />);
    expect(container).toMatchSnapshot();

    eventEmitter.emit('UPDATE_DPI', 'high');
    await new Promise((resolve) => setTimeout(resolve));
    expect(container).toMatchSnapshot();
  });

  it('should render correctly in mobile', () => {
    mockUseIsMobile.mockReturnValue(true);
    mockRead.mockReturnValue('ultra');
    const { container } = render(<DpiInfo />);
    expect(container).toMatchSnapshot();
  });
});
