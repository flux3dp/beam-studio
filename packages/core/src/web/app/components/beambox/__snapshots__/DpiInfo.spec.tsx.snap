// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test DpiInfo should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <i>
      Resolution
      : 
      125
       DPI
    </i>
  </div>
</div>
`;

exports[`test DpiInfo should render correctly 2`] = `
<div>
  <div
    class="container"
  >
    <i>
      Resolution
      : 
      500
       DPI
    </i>
  </div>
</div>
`;

exports[`test DpiInfo should render correctly in mobile 1`] = `
<div>
  <div
    class="container mobile"
  >
    <i>
      Resolution
      : 
      1000
       DPI
    </i>
  </div>
</div>
`;
